### DISCUSSION TITLE
keypoint detection & heatmap, difference with paper

### DISCUSSION THREAD

**Discussion by joihn** [facebookresearch/detectron2#3794] [Q&A]
I'm studying the way keypoint detection is implemented in detectron2 compared to maskrcnn paper.

maskrcnn authors: "we minimize the cross-entropy loss over an m2 -way softmax output (which encourages a single point to be detected)"

but it looks like in detectron2 :

the loss is a simple cross entropy over the mask:
https://github.com/facebookresearch/detectron2/blob/main/detectron2/modeling/roi_heads/keypoint_head.py#L88

And then the softmax is applied **later** only to get the keypoint score:
https://github.com/facebookresearch/detectron2/blob/main/detectron2/structures/keypoints.py#L218
Then there is slight discrepency with the paper, Am I correct ? Is there a reason ? Maybe it works better that way ?

**Comment by ppwwyyxx** [comment]
> the loss is a simple cross entropy over the mask:

Pytorch's cross_entropy function already contains a softmax itself: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss
